presto-spi/src/main/java/com/facebook/presto/spi/block/ArrayElementBlockWriter.java

```diff
@@ -0,0 +1,145 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.block;
+
+import io.airlift.slice.Slice;
+
+public class ArrayElementBlockWriter
+        extends AbstractArrayElementBlock
+        implements BlockBuilder
+{
+    private BlockBuilder blockBuilder;
+    private int positionsWritten;
+
+    public ArrayElementBlockWriter(BlockBuilder blockBuilder, int start)
+    {
+        super(start);
+        this.blockBuilder = blockBuilder;
+    }
+
+    @Override
+    protected BlockBuilder getBlock()
+    {
+        return blockBuilder;
+    }
+
+    @Override
+    public BlockBuilder writeByte(int value)
+    {
+        blockBuilder.writeByte(value);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder writeShort(int value)
+    {
+        blockBuilder.writeShort(value);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder writeInt(int value)
+    {
+        blockBuilder.writeInt(value);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder writeLong(long value)
+    {
+        blockBuilder.writeLong(value);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder writeFloat(float value)
+    {
+        blockBuilder.writeFloat(value);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder writeDouble(double value)
+    {
+        blockBuilder.writeDouble(value);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder writeBytes(Slice source, int sourceIndex, int length)
+    {
+        blockBuilder.writeBytes(source, sourceIndex, length);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder writeObject(Object value)
+    {
+        blockBuilder.writeObject(value);
+        return this;
+    }
+
+    @Override
+    public BlockBuilder nextObjectWriter()
+    {
+        return blockBuilder.nextObjectWriter();
+    }
+
+    @Override
+    public BlockBuilder appendNull()
+    {
+        blockBuilder.appendNull();
+        entryAdded();
+        return this;
+    }
+
+    @Override
+    public BlockBuilder closeEntry()
+    {
+        blockBuilder.closeEntry();
+        entryAdded();
+        return this;
+    }
+
+    private void entryAdded()
+    {
+        positionsWritten++;
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return positionsWritten;
+    }
+
+    @Override
+    public Block build()
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    public void closeWriter(BlockBuilder creator)
```

Comment:
Instead of making the caller of `nextObjectWriter` call this method, I would store the current writer into a member variable in `ArrayBlockBuilder.nextObjectWriter`, return `this` from `ArrayElementWriter.build()` and in `ArrayBlockBuilder.writeObject` check if the `value == currentWriter`. Then you can remove this method, and the workflow for callers will be very natural:
```
BlockBlocker array = builder.nextObjectWriter();
// write stuff to array
builder.writeObject(array.build())
```